### ISSUE TITLE
SSL connect errors when testing AWS API Gateway endpoint

### ISSUE THREAD

**Issue by coopernurse** [wg/wrk#149]
This may not be a wrk problem, but if you kind folks have a moment perhaps you can explain what I'm seeing.

I'm testing an AWS Lambda function via AWS API Gateway.  I've tried using wrk via home brew and by compiling manually using the latest master sources (OS X 10.10.4).  I simply get socket errors:

```
$ wrk -d 2s -s helloLambda.lua https://91lzn61el4.execute-api.us-west-2.amazonaws.com/prod/hello-world
Running 2s test @ https://91lzn61el4.execute-api.us-west-2.amazonaws.com/prod/hello-world
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     0.00us    0.00us   0.00us     nan%
    Req/Sec     0.00      0.00     0.00       nan%
  0 requests in 2.10s, 0.00B read
  Socket errors: connect 376, read 0, write 0, timeout 0
Requests/sec:      0.00
Transfer/sec:       0.00B
```

curl is able to successfully POST to this URL:
```
$ curl -v -X POST -d '{ "key1" : "blah", "key2" : "baz" }' https://91lzn61el4.execute-api.us-west-2.amazonaws.com/prod/hello-world
* Hostname was NOT found in DNS cache
*   Trying 54.192.68.188...
* Connected to 91lzn61el4.execute-api.us-west-2.amazonaws.com (54.192.68.188) port 443 (#0)
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256
* Server certificate: *.execute-api.us-west-2.amazonaws.com
* Server certificate: Symantec Class 3 Secure Server CA - G4
* Server certificate: VeriSign Class 3 Public Primary Certification Authority - G5
> POST /prod/hello-world HTTP/1.1
> User-Agent: curl/7.37.1
> Host: 91lzn61el4.execute-api.us-west-2.amazonaws.com
> Accept: */*
> Content-Length: 35
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 35 out of 35 bytes
< HTTP/1.1 200 OK
< Content-Type: application/json
< Content-Length: 6
< Connection: keep-alive
< Date: Sat, 11 Jul 2015 20:24:53 GMT
< x-amzn-RequestId: e3696fcb-280a-11e5-a22c-0930742c392a
< X-Cache: Miss from cloudfront
< Via: 1.1 7a0918c01bce16cc9b165fd895f7dc87.cloudfront.net (CloudFront)
< X-Amz-Cf-Id: TEnh3fD0hTi_7tFP2t-iuSBbvHrY_BLadzzHYwshrcIDf1fy7cktJw==
< 
* Connection #0 to host 91lzn61el4.execute-api.us-west-2.amazonaws.com left intact
"blah"
```

Any suggestions on how to debug this?  I'm not sure if #117 would aid this case.  For what it's worth I'm not having much luck with `ab` either.  Only curl seems to negotiate properly with this endpoint.

Thanks in advance.

**Comment by coopernurse** [comment]
Here's the `helloLambda.lua` file I used in the above invocation:

```
wrk.method = "POST"
wrk.body   = '{ "key1" : "blah", "key2" : "baz" }'
wrk.headers["Content-Type"] = "application/json"
```

**Comment by wg** [comment]
Hey @coopernurse, that's definitely a weird problem. Have you tried specifying an IP address instead of the hostname? Looks like it resolves to a bunch of different IPs, I wonder if some don't respond properly? Another option might be seeing if it works without TLS.

**Comment by coopernurse** [comment]
Thanks for following up.  

Using standard http (not https) results in a 403 error code.  

Using an IP address doesn't appear to work.  Same connect errors with wrk. curl also fails by IP with: `(35) SSL peer handshake failed, the server most likely requires a client certificate to connect` - this occurs even with the -k option.

Tools that don't seem to work:
* wrk
* ab
* siege

Tools that can connect:
* curl
* gobench (https://github.com/cmpxchg16/gobench)

I've posted a thread here:

https://forums.aws.amazon.com/thread.jspa?threadID=193615&tstart=25

The only reply so far suggests using TLS 1.2 to negotiate.  #117 would probably be relevant here.

**Comment by wg** [comment]
Thanks for investigating further @coopernurse, this was a tricky one to figure out. Turns out that whatever AWS is using to terminate TLS requires SNI which wrk doesn't support. I'll add support soon but in the meantime this patch should get you up and running:

```
diff --git a/src/ssl.c b/src/ssl.c
index 604bf0b..21c0ca6 100644
--- a/src/ssl.c
+++ b/src/ssl.c
@@ -52,6 +52,7 @@ SSL_CTX *ssl_init() {
 status ssl_connect(connection *c) {
     int r;
     SSL_set_fd(c->ssl, c->fd);
+    SSL_set_tlsext_host_name(c->ssl, "91lzn61el4.execute-api.us-west-2.amazonaws.com");
     if ((r = SSL_connect(c->ssl)) != 1) {
         switch (SSL_get_error(c->ssl, r)) {
             case SSL_ERROR_WANT_READ:  return RETRY;
```

**Comment by coopernurse** [comment]
Wow, thanks very much for digging into this.  Much appreciated.

**Comment by Govinda-Fichtner** [comment]
Is the support for sni still coming?

**Comment by bubenkoff** [comment]
+1 for solving this

**Comment by wg** [comment]
Thanks for your patience, I've finally released wrk 4.0.2 which supports SNI. Doing so requires a more recent version of OpenSSL than some platforms provide, so I've bundled OpenSSL 1.0.2 into the wrk source distribution.
